ddtrace/tracer/textmap.go

```diff
@@ -264,46 +265,68 @@ func (p *chainedPropagator) Inject(spanCtx ddtrace.SpanContext, carrier interfac
 
 // Extract implements Propagator. This method will attempt to extract the context
 // based on the precedence order of the propagators. Generally, the first valid
-// trace context that could be extracted will be returned, and other extractors will
-// be ignored. However, the W3C tracestate header value will always be extracted and
-// stored in the local trace context even if a previous propagator has already succeeded
-// so long as the trace-ids match.
+// trace context that could be extracted will be returned. However, the W3C tracestate
+// header value will always be extracted and stored in the local trace context even if
+// a previous propagator has already succeeded so long as the trace-ids match.
+// Furthermore, if we have already successfully extracted a trace context and a
+// subsequent trace context has conflicting trace information, such information will
+// be relayed in the original trace context with a SpanLink.
 func (p *chainedPropagator) Extract(carrier interface{}) (ddtrace.SpanContext, error) {
-	var ctx ddtrace.SpanContext
+	var ctx ddtrace.SpanContext // First valid trace context will be stored here
+	var links []ddtrace.SpanLink
 	for _, v := range p.extractors {
-		if ctx != nil {
-			// A local trace context has already been extracted.
-			pw3c, isW3C := v.(*propagatorW3c)
-			if !isW3C {
-				continue // Ignore other propagators.
-			}
-			w3cCtx, err := pw3c.Extract(carrier)
-			if err == nil && w3cCtx.(*spanContext).TraceID128() == ctx.(*spanContext).TraceID128() {
-				pw3c.propagateTracestate(ctx.(*spanContext), w3cCtx.(*spanContext))
-				if w3cCtx.SpanID() != ctx.SpanID() {
-					var ddCtx *spanContext
-					if ddp := getDatadogPropagator(p); ddp != nil {
-						if ddSpanCtx, err := ddp.Extract(carrier); err == nil {
-							ddCtx, _ = ddSpanCtx.(*spanContext)
+		extractedCtx, err := v.Extract(carrier)
+		if ctx != nil { // A local trace context has already been extracted
+			if err == nil {
+				propagatorType, unkPropagatorType := getPropagatorType(v)
+				if unkPropagatorType != nil { // If the propagator type is not one that we expect, ignore this extraction
+					continue
+				}
+				if extractedCtx.(*spanContext).TraceID128() == ctx.(*spanContext).TraceID128() {
```

Comment:
I know you just reused old code at this line but I am wondering why SpanContextW3C isn't used here instead.